File: src/main/java/com/freya02/botcommands/api/application/slash/annotations/VarArgs.java
package com.freya02.botcommands.api.application.slash.annotations;

import com.freya02.botcommands.api.application.annotations.AppOption;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.List;

//TODO see if we can auto generate description based off a MessageFormat, could cause issues with localization
/**
 * Allows generating N command options from the specified {@link AppOption}
 * <br>The target parameter must be of type {@link List}
 * <br>You can configure how many arguments are required with {@link #numRequired()}
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.PARAMETER)
public @interface VarArgs {
	/**
	 * @return The number of times this option needs to appear, must be between 1 and 25 (max number of options)
	 */
	int value();

	/**
	 * @return The number of required options for this vararg
	 */
	int numRequired() default 1;
}

File: src/main/kotlin/io/github/freya022/botcommands/api/commands/annotations/VarArgs.kt
package io.github.freya022.botcommands.api.commands.annotations

import io.github.freya022.botcommands.api.commands.application.slash.annotations.SlashOption
import io.github.freya022.botcommands.api.commands.application.slash.builder.SlashCommandBuilder
import io.github.freya022.botcommands.api.commands.text.annotations.TextOption
import io.github.freya022.botcommands.api.commands.text.builder.TextCommandVariationBuilder

/**
 * Generates N command options from the specified [@SlashOption][SlashOption] or [@TextOption][TextOption].
 *
 * The target parameter must be of type [List].
 *
 * You can configure how many arguments are required with [numRequired].
 *
 * See [@MentionsString][MentionsString] for a way to get a list of mentionable (user/member/role/channel...)
 * without a vararg, using a single string.
 *
 * **Note:** You are limited to one vararg parameter in text commands.
 *
 * @see TextCommandVariationBuilder.optionVararg
 * @see TextCommandVariationBuilder.inlineClassOptionVararg
 *
 * @see SlashCommandBuilder.optionVararg
 * @see SlashCommandBuilder.inlineClassOptionVararg
 *
 * @see MentionsString @MentionsString
 */
@Target(AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
annotation class VarArgs(
    /**
     * The number of times this option needs to appear, which must be between 1 and {@value CommandData#MAX_OPTIONS}.
     */
    val value: Int,

    /**
     * The number of required options for this vararg.
     */
    val numRequired: Int = 1
)
